File: _system/trading/portfolio_hub/tests/test_command_peek.py
"""The bridge peeks before it claims, so an idle desk writes nothing to D1.

Each claim route reserves a nonce row before it looks for work. The loop used to
call both claim routes every 15s tick, so an idle desk wrote two rows every tick
(~34k billed row writes a day, plus ~10k more when retention deleted them). These
tests pin the new contract: claims happen only for a queue the peek says has
claimable work, and every way the peek can be missing or broken falls back to
the old claim-both behaviour rather than stranding a ticket.
"""
from __future__ import annotations

import hashlib
import hmac
import io
import json
import urllib.error

import pytest

from _system.trading.portfolio_hub import command_poller
from _system.trading.portfolio_hub.command_poller import (
    IDLE_POLL_SECONDS,
    PEEK_PATH,
    PEEK_RETRY_SECONDS,
    ChannelConfig,
    OrderCommandChannel,
    OrderCommandLoop,
    PeekUnavailable,
)
from _system.trading.portfolio_hub.paper import PaperOrderBroker


def _quotes(conid):
    return {"bid": "5.10", "ask": "5.20", "min_tick": "0.01", "multiplier": "100"}


class _Service:
    def __init__(self, broker):
        self.broker = broker


class _PeekingChannel:
    """Counts signed calls. Each claim is one nonce row at the edge; a peek is none."""

    def __init__(self, *, peek=None, requests=None, lookups=None):
        self._peek = peek if peek is not None else {"order_requests": 0, "contract_lookups": 0}
        self._requests, self._lookups = requests or [], lookups or []
        self.calls: list[str] = []
        self.published, self.lookup_published = [], []

    def peek(self):
        self.calls.append("peek")
        if isinstance(self._peek, BaseException):
            raise self._peek
        return dict(self._peek)

    def claim(self):
        self.calls.append("claim")
        return self._requests

    def claim_lookups(self):
        self.calls.append("claim_lookups")
        return self._lookups

    def publish(self, request_id, update):
        self.published.append((request_id, update))

    def publish_lookup(self, lookup_id, update):
        self.lookup_published.append((lookup_id, update))


class _Clock:
    def __init__(self):
        self.now = 1_000.0

    def __call__(self):
        return self.now


def _loop(channel, clock=None):
    return OrderCommandLoop(_Service(PaperOrderBroker(_quotes)), channel, account_alias="U123",
                            clock=clock or _Clock())


def _nonce_writes(channel):
    # Each claim route reserves exactly one nonce; the peek reserves none.
    return sum(1 for call in channel.calls if call in {"claim", "claim_lookups"})


def test_an_idle_desk_makes_no_claim_and_so_writes_no_nonce():
    channel = _PeekingChannel()
    loop = _loop(channel)
    for _ in range(100):
        assert loop.tick() is False
    assert _nonce_writes(channel) == 0, "100 idle ticks must reserve zero nonces"
    assert channel.calls == ["peek"] * 100


def test_only_the_queue_with_claimable_work_is_claimed():
    orders_only = _PeekingChannel(peek={"order_requests": 1, "contract_lookups": 0})
    _loop(orders_only).tick()
    assert orders_only.calls == ["peek", "claim"]

    lookups_only = _PeekingChannel(peek={"order_requests": 0, "contract_lookups": 2})
    _loop(lookups_only).tick()
    assert lookups_only.calls == ["peek", "claim_lookups"]


def test_work_found_through_the_peek_is_handled_exactly_as_before():
    """Same ticket, same refusal: the peek changes when we claim, never what we do."""
    row = {
        "request_id": "r1", "conid": 907480285, "sec_type": "OPT", "action": "BUY",
        "quantity_decimal": "2", "limit_price_decimal": "5.15", "owner": "drew",
        "strategy": "single_stock", "mode": "paper", "tif": "DAY", "state": "requested",
    }
    channel = _PeekingChannel(peek={"order_requests": 1, "contract_lookups": 0}, requests=[row])
    _loop(channel).tick()
    assert channel.published and channel.published[0][1]["state"] == "rejected"
    assert "options interlock" in channel.published[0][1]["reject_reason"]


def test_a_missing_peek_route_falls_back_to_claiming_and_reprobes_later():
    """The edge may deploy after the bridge; nothing may be stranded meanwhile."""
    clock = _Clock()
    channel = _PeekingChannel(peek=PeekUnavailable("peek route answered HTTP 405"))
    loop = _loop(channel, clock)

    loop.tick()
    assert channel.calls == ["peek", "claim", "claim_lookups"], "fall back to the old claim-both tick"

    channel.calls.clear()
    clock.now += PEEK_RETRY_SECONDS - 1
    loop.tick()
    assert channel.calls == ["claim", "claim_lookups"], "no re-probe inside the retry window"

    channel.calls.clear()
    channel._peek = {"order_requests": 0, "contract_lookups": 0}
    clock.now += 2
    loop.tick()
    assert channel.calls == ["peek"], "once the route exists the saving starts on its own"


def test_a_failing_peek_claims_directly_for_that_tick_only():
    channel = _PeekingChannel(peek=RuntimeError("HTTP 500 from the edge"))
    loop = _loop(channel)
    loop.tick()
    assert channel.calls == ["peek", "claim", "claim_lookups"]
    channel.calls.clear()
    channel._peek = {"order_requests": 0, "contract_lookups": 0}
    loop.tick()
    assert channel.calls == ["peek"], "a transient failure does not suspend the peek"


def test_a_channel_without_a_peek_behaves_as_it_always_did():
    class _Legacy:
        def __init__(self):
            self.calls = []

        def claim(self):
            self.calls.append("claim")
            return []

        def claim_lookups(self):
            self.calls.append("claim_lookups")
            return []

    channel = _Legacy()
    _loop(channel).tick()
    assert channel.calls == ["claim", "claim_lookups"]


def test_run_forever_on_an_idle_desk_sleeps_idle_and_never_claims():
    channel = _PeekingChannel()
    slept: list[float] = []

    def sleep(seconds):
        slept.append(seconds)
        if len(slept) == 5:
            raise KeyboardInterrupt

    with pytest.raises(KeyboardInterrupt):
        _loop(channel).run_forever(sleep=sleep)
    assert slept == [IDLE_POLL_SECONDS] * 5
    assert _nonce_writes(channel) == 0


# ---------------------------------------------------------------- the channel

class _Response(io.BytesIO):
    def __enter__(self):
        return self

    def __exit__(self, *exc):
        return False


def _channel():
    return OrderCommandChannel(ChannelConfig(base_url="https://dash.example", token="t" * 40,
                                             account_alias="U123"))


def test_peek_is_a_signed_post_to_the_peek_route(monkeypatch):
    seen = {}

    def urlopen(request, timeout):
        seen["url"], seen["method"], seen["body"] = request.full_url, request.get_method(), request.data
        seen["headers"] = {key.lower(): value for key, value in request.header_items()}
        return _Response(json.dumps({"claimable": {"order_requests": 3, "contract_lookups": 0}}).encode())

    monkeypatch.setattr(command_poller.urllib.request, "urlopen", urlopen)
    assert _channel().peek() == {"order_requests": 3, "contract_lookups": 0}
    assert seen["url"] == f"https://dash.example{PEEK_PATH}"
    assert seen["method"] == "POST", "a GET would be answered by the SPA fallback page before deploy"
    assert json.loads(seen["body"]) == {"account_alias": "U123"}
    headers = seen["headers"]
    expected = hmac.new(("t" * 40).encode(),
                        f"{headers['x-portfolio-timestamp']}\n{headers['x-portfolio-nonce']}\n".encode() + seen["body"],
                        hashlib.sha256).hexdigest()
    assert headers["x-portfolio-signature"] == expected


@pytest.mark.parametrize("code", [404, 405])
def test_an_undeployed_peek_route_is_reported_as_unavailable(monkeypatch, code):
    def urlopen(request, timeout):
        raise urllib.error.HTTPError(request.full_url, code, "no route", {}, io.BytesIO(b""))

    monkeypatch.setattr(command_poller.urllib.request, "urlopen", urlopen)
    with pytest.raises(PeekUnavailable):
        _channel().peek()


def test_a_pages_fallback_page_is_reported_as_unavailable(monkeypatch):
    monkeypatch.setattr(command_poller.urllib.request, "urlopen",
                        lambda request, timeout: _Response(b"<!doctype html><title>dashboard</title>"))
    with pytest.raises(PeekUnavailable):
        _channel().peek()


def test_a_server_error_is_not_mistaken_for_a_missing_route(monkeypatch):
    def urlopen(request, timeout):
        raise urllib.error.HTTPError(request.full_url, 500, "boom", {}, io.BytesIO(b""))

    monkeypatch.setattr(command_poller.urllib.request, "urlopen", urlopen)
    with pytest.raises(urllib.error.HTTPError):
        _channel().peek()

File: dashboard/cloudflare/test-command-peek.mjs
// The command-channel peek, against real SQL.
//
// Two properties, and D1 bills for the first while the second is security:
//
//   * An idle bridge tick writes nothing. Before the peek, every 15s tick called
//     both claim routes, and each reserved a nonce row first -- ~34k billed row
//     writes a day, plus ~10k more when retention deleted them.
//   * A peek authenticates as a peek and as nothing else. It reserves no nonce,
//     so a peek that verified on any other route would be a fresh, unused
//     signature there for its whole 300s window. The first version shipped
//     exactly that: a captured peek replayed to a claim route leased the desk's
//     tickets away from the real bridge.

import assert from "node:assert/strict";
import test from "node:test";
import { DatabaseSync } from "node:sqlite";
import { readFileSync, readdirSync } from "node:fs";
import { createHmac, randomUUID } from "node:crypto";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";

import { onRequestPost as peek, PEEK_CONTRACT_LOOKUPS_SQL, PEEK_ORDER_REQUESTS_SQL } from "../functions/api/v2/portfolio/ingest/peek.js";
import { onRequestPost as claimOrders } from "../functions/api/v2/portfolio/ingest/order-requests/claim.js";
import { onRequestPost as claimLookups } from "../functions/api/v2/portfolio/ingest/contract-lookups/claim.js";
import { onRequestPost as publishOrder } from "../functions/api/v2/portfolio/ingest/order-requests/publish.js";
import { onRequestPost as publishLookup } from "../functions/api/v2/portfolio/ingest/contract-lookups/publish.js";
import { onRequestPost as ingest } from "../functions/api/v2/portfolio/ingest.js";

const HERE = dirname(fileURLToPath(import.meta.url));
const TOKEN = "x".repeat(48);
const ACCOUNT = "U805366";

// A D1-shaped adapter whose batch() returns per-statement results the way D1
// does, so a batched SELECT is read back exactly as production reads it.
function d1(database) {
  const make = (sql, binds = []) => {
    const values = () => binds.map((value) => (value === undefined ? null : typeof value === "boolean" ? Number(value) : value));
    return {
      sql,
      bind: (...next) => make(sql, next),
      run: async () => ({ meta: { changes: Number(database.prepare(sql).run(...values()).changes || 0) } }),
      all: async () => ({ results: database.prepare(sql).all(...values()) }),
      first: async () => database.prepare(sql).get(...values()) ?? null,
      _batch: () => (/^\s*select/i.test(sql)
        ? { results: database.prepare(sql).all(...values()) }
        : { results: [], meta: { changes: Number(database.prepare(sql).run(...values()).changes || 0) } }),
    };
  };
  return { prepare: (sql) => make(sql), batch: async (statements) => statements.map((statement) => statement._batch()) };
}

function freshDatabase() {
  const database = new DatabaseSync(":memory:");
  const dir = join(HERE, "migrations");
  for (const file of readdirSync(dir).filter((name) => name.endsWith(".sql")).sort()) {
    const sql = readFileSync(join(dir, file), "utf8").replace(/--[^\n]*/g, "");
    for (const statement of sql.split(";").map((part) => part.trim()).filter(Boolean)) {
      try { database.exec(statement); } catch (_) { /* unrelated legacy tables */ }
    }
  }
  return database;
}

/**
 * A signed request exactly as publisher.signed_headers builds it: the body is
 * the canonical JSON (sorted keys, no spaces) and, with a domain, the signed
 * message is "<domain>\n<ts>\n<nonce>\n<body>".
 */
function signed(payload, { domain = null, url = "https://dash.example/x" } = {}) {
  const body = JSON.stringify(Object.fromEntries(Object.entries(payload).sort(([a], [b]) => a.localeCompare(b))));
  const timestamp = String(Math.floor(Date.now() / 1000));
  const nonce = randomUUID().replace(/-/g, "");
  const signature = createHmac("sha256", TOKEN).update(`${domain ? `${domain}\n` : ""}${timestamp}\n${nonce}\n${body}`).digest("hex");
  return new Request(url, {
    method: "POST", body,
    headers: {
      "content-type": "application/json", "x-portfolio-timestamp": timestamp,
      "x-portfolio-nonce": nonce, "x-portfolio-signature": signature,
    },
  });
}

// What the new bridge sends (command_poller.py): a peek-domain signature over a
// body that names its purpose, and claims that say purpose "claim".
const peekRequest = () => signed({ account_alias: ACCOUNT, purpose: "peek" }, { domain: "peek" });
const env = (database) => ({ DB: d1(database), PORTFOLIO_INGEST_TOKEN: TOKEN, PRIVATE_ARTIFACTS: { put: async () => {} } });

async function send(handler, database, request) {
  const response = await handler({ request, env: env(database) });
  return { status: response.status, body: await response.json() };
}

const totalChanges = (database) => database.prepare("SELECT total_changes() AS n").get().n;
const nonces = (database) => database.prepare("SELECT COUNT(*) AS n FROM portfolio_ingest_nonces").get().n;
const ago = (seconds) => new Date(Date.now() - seconds * 1000).toISOString();

function seedRequest(database, overrides = {}) {
  const row = {
    request_id: randomUUID(), account_alias: ACCOUNT, owner: "drew", strategy: "single_stock",
    conid: 272093, symbol: "MSFT", sec_type: "STK", action: "BUY", quantity_decimal: "1",
    limit_price_decimal: "400", tif: "DAY", outside_rth: 0, mode: "paper", state: "requested",
    created_at: ago(5), updated_at: ago(5), ...overrides,
  };
  database.prepare(`INSERT INTO portfolio_order_requests (${Object.keys(row).join(",")})
    VALUES (${Object.keys(row).map(() => "?").join(",")})`).run(...Object.values(row));
  return row.request_id;
}

function seedLookup(database, overrides = {}) {
  const row = {
    lookup_id: randomUUID(), account_alias: ACCOUNT, owner: "drew", kind: "contract",
    symbol: "MSFT", sec_type: "STK", state: "requested", created_at: ago(5), updated_at: ago(5), ...overrides,
  };
  database.prepare(`INSERT INTO portfolio_contract_lookups (${Object.keys(row).join(",")})
    VALUES (${Object.keys(row).map(() => "?").join(",")})`).run(...Object.values(row));
  return row.lookup_id;
}

const leased = (database) => ({
  orders: database.prepare("SELECT COUNT(*) AS n FROM portfolio_order_requests WHERE claimed_at IS NOT NULL").get().n,
  lookups: database.prepare("SELECT COUNT(*) AS n FROM portfolio_contract_lookups WHERE claimed_at IS NOT NULL").get().n,
});

// ------------------------------------------------------------- idle cost

test("an idle peek writes no row at all -- no nonce, no lease", async () => {
  const database = freshDatabase();
  const before = totalChanges(database);
  const { status, body } = await send(peek, database, peekRequest());
  assert.equal(status, 200);
  assert.deepEqual(body.claimable, { order_requests: 0, contract_lookups: 0 });
  assert.equal(totalChanges(database) - before, 0, "the idle peek must not write");
  assert.equal(nonces(database), 0);
});

test("for contrast: the claims it replaces write a nonce row each, idle or not", async () => {
  const database = freshDatabase();
  await send(claimOrders, database, signed({ account_alias: ACCOUNT, purpose: "claim" }));
  await send(claimLookups, database, signed({ account_alias: ACCOUNT, purpose: "claim" }));
  assert.equal(nonces(database), 2);
});

test("a replayed peek is answered as a peek and still changes nothing", async () => {
  const database = freshDatabase();
  const key = seedRequest(database);
  const request = peekRequest();
  const before = totalChanges(database);
  assert.equal((await peek({ request: request.clone(), env: env(database) })).status, 200);
  assert.equal((await peek({ request, env: env(database) })).status, 200, "no nonce: a replay inside the window is answered");
  assert.equal(totalChanges(database) - before, 0);
  const row = database.prepare("SELECT state, claimed_at FROM portfolio_order_requests WHERE request_id=?").get(key);
  assert.equal(row.state, "requested");
  assert.equal(row.claimed_at, null, "a peek, replayed or not, never takes a lease");
});

// --------------------------------------------- a peek is never anything else

test("no request the peek accepts is accepted by any other signed route", async () => {
  // Every shape a bridge has ever sent to the peek route. Whichever of them the
  // peek accepts, replaying that exact request anywhere else must be refused
  // and must lease, move or write nothing. (The first shape is what the first
  // version of the bridge sent; against that edge this test fails.)
  const candidates = {
    "legacy signature, bare account body": () => signed({ account_alias: ACCOUNT }),
    "legacy signature, peek-purpose body": () => signed({ account_alias: ACCOUNT, purpose: "peek" }),
    "peek-domain signature, peek-purpose body": peekRequest,
  };
  const elsewhere = { claimOrders, claimLookups, publishOrder, publishLookup, ingest };
  let accepted = 0;
  for (const [shape, build] of Object.entries(candidates)) {
    const database = freshDatabase();
    const ticket = seedRequest(database);
    seedLookup(database);
    const request = build();
    const answer = await peek({ request: request.clone(), env: env(database) });
    if (answer.status !== 200) continue;
    accepted += 1;
    const before = totalChanges(database);
    for (const [route, handler] of Object.entries(elsewhere)) {
      const replay = await handler({ request: request.clone(), env: env(database) });
      assert.ok(replay.status >= 400 && replay.status < 500, `${shape} replayed to ${route} answered ${replay.status}`);
    }
    assert.deepEqual(leased(database), { orders: 0, lookups: 0 }, `${shape}: a replay leased work away from the bridge`);
    assert.equal(totalChanges(database) - before, 0, `${shape}: a replay wrote to D1`);
    assert.equal(database.prepare("SELECT state FROM portfolio_order_requests WHERE request_id=?").get(ticket).state, "requested");
  }
  assert.ok(accepted >= 1, "the peek must accept the bridge's own peek, or this test proves nothing");
});

test("both claim routes refuse a peek-purpose body even when signed the claim way, before any write", async () => {
  // The second, independent lock: a body that says it is a peek is never a
  // claim, whatever signature it carries.
  const database = freshDatabase();
  seedRequest(database);
  seedLookup(database);
  const before = totalChanges(database);
  for (const handler of [claimOrders, claimLookups]) {
    const { status, body } = await send(handler, database, signed({ account_alias: ACCOUNT, purpose: "peek" }));
    assert.equal(status, 422);
    assert.match(body.error, /not a claim/);
  }
  assert.deepEqual(leased(database), { orders: 0, lookups: 0 });
  assert.equal(totalChanges(database) - before, 0, "refused before the nonce was reserved");
});

test("the deployed bridge's legacy claim body (no purpose) still claims, as does purpose: claim", async () => {
  for (const payload of [{ account_alias: ACCOUNT }, { account_alias: ACCOUNT, purpose: "claim" }]) {
    const database = freshDatabase();
    const ticket = seedRequest(database);
    seedLookup(database);
    const orders = await send(claimOrders, database, signed(payload));
    const lookups = await send(claimLookups, database, signed(payload));
    assert.equal(orders.status, 200);
    assert.equal(orders.body.requests[0].request_id, ticket);
    assert.equal(lookups.status, 200);
    assert.equal(lookups.body.lookups.length, 1);
  }
});

test("the peek refuses a claim-style signature and a body without its purpose", async () => {
  const database = freshDatabase();
  const claimSigned = await send(peek, database, signed({ account_alias: ACCOUNT, purpose: "peek" }));
  assert.equal(claimSigned.status, 401, "a claim or ingest signature is not a peek signature");
  const unlabelled = await send(peek, database, signed({ account_alias: ACCOUNT }, { domain: "peek" }));
  assert.equal(unlabelled.status, 422, "the signed body must say what it is");
});

test("an unsigned or tampered peek is refused", async () => {
  const database = freshDatabase();
  const unsigned = await peek({ request: new Request("https://dash.example/x", { method: "POST", body: "{}" }), env: env(database) });
  assert.equal(unsigned.status, 401);
  const request = peekRequest();
  const tampered = new Request(request.url, { method: "POST", headers: request.headers, body: JSON.stringify({ account_alias: "U999", purpose: "peek" }) });
  assert.equal((await peek({ request: tampered, env: env(database) })).status, 401);
});

// ------------------------------------------------------------- semantics

test("the peek counts exactly what a claim would take", async () => {
  const database = freshDatabase();
  seedRequest(database);                                                   // claimable
  seedRequest(database, { state: "previewed", claimed_at: ago(10), claimed_by: "b" }); // lease live
  seedRequest(database, { state: "approved", claimed_at: ago(120), claimed_by: "b" }); // lease lapsed
  seedRequest(database, { state: "filled" });                              // terminal
  seedRequest(database, { account_alias: "OTHER" });                       // other account
  seedLookup(database);                                                    // claimable
  seedLookup(database, { state: "resolving", claimed_at: ago(5), claimed_by: "b" });   // lease live
  seedLookup(database, { state: "resolved" });                             // done

  const { body } = await send(peek, database, peekRequest());
  assert.deepEqual(body.claimable, { order_requests: 2, contract_lookups: 1 });

  const orders = await send(claimOrders, database, signed({ account_alias: ACCOUNT, purpose: "claim" }));
  const lookups = await send(claimLookups, database, signed({ account_alias: ACCOUNT, purpose: "claim" }));
  assert.equal(orders.body.requests.length, body.claimable.order_requests);
  assert.equal(lookups.body.lookups.length, body.claimable.contract_lookups);

  // And once claimed, the peek goes quiet for the length of the lease.
  const after = await send(peek, database, peekRequest());
  assert.deepEqual(after.body.claimable, { order_requests: 0, contract_lookups: 0 });
});

test("both peek statements seek an index instead of scanning the table", () => {
  const database = freshDatabase();
  const plan = (sql, binds) => database.prepare(`EXPLAIN QUERY PLAN ${sql}`).all(...binds).map((row) => row.detail).join(" | ");
  const orders = plan(PEEK_ORDER_REQUESTS_SQL, ["requested", "drafting", "previewed", "approved", "submitting", ACCOUNT, ago(90)]);
  const lookups = plan(PEEK_CONTRACT_LOOKUPS_SQL, ["requested", "resolving", ACCOUNT, ago(60)]);
  assert.match(orders, /USING (COVERING )?INDEX idx_order_requests_state/, orders);
  assert.match(lookups, /USING (COVERING )?INDEX idx_contract_lookups_state/, lookups);
  assert.doesNotMatch(`${orders} ${lookups}`, /SCAN portfolio_/);
});

test("the peek requires an account", async () => {
  const database = freshDatabase();
  assert.equal((await send(peek, database, signed({ purpose: "peek" }, { domain: "peek" }))).status, 422);
});
